platform/kube/inject/initializer.go

```diff
@@ -0,0 +1,456 @@
+	}
+	return i
+}
+
+func (i *Initializer) initializerPresent(objectMeta *metav1.ObjectMeta) bool {
+	glog.V(2).Infof("ObjectMeta initializer info %v/%v policy:%q status:%q %v",
+		objectMeta.Namespace, objectMeta.Name,
+		objectMeta.Annotations[istioSidecarAnnotationPolicyKey],
+		objectMeta.Annotations[istioSidecarAnnotationStatusKey],
+		objectMeta.Initializers)
+
+	if objectMeta.GetInitializers() == nil {
+		return false
+	}
+	pendingInitializers := objectMeta.GetInitializers().Pending
+	if len(pendingInitializers) == 0 {
+		return false
+	}
+	if initializerName != pendingInitializers[0].Name {
+		return false
+	}
+
+	return true
+}
+
+func (i *Initializer) modifyResource(objectMeta *metav1.ObjectMeta, templateObjectMeta *metav1.ObjectMeta, spec *v1.PodSpec) error { // nolint: lll
+	switch i.options.Namespace {
+	case v1.NamespaceAll:
+		// skip kubernetes system namespace
+		if objectMeta.Namespace == "kube-system" {
+			return nil
+		}
+	case objectMeta.Namespace:
+	default:
+		// Skip namespace(s) that we're not responsible for
+		// initializing.
+		return nil
+	}
+
+	// Remove self from the list of pending Initializers while
+	// preserving ordering.
+	if pending := objectMeta.GetInitializers().Pending; len(pending) == 1 {
+		objectMeta.Initializers = nil
+	} else {
+		objectMeta.Initializers.Pending = append(pending[:0], pending[1:]...)
+	}
+
+	if !injectRequired(i.options.InjectionPolicy, objectMeta) {
+		glog.V(2).Infof("Skipping %s/%s", objectMeta.Namespace, objectMeta.Name)
+		return nil
+	}
+
+	glog.Infof("Initializing %s/%s", objectMeta.Namespace, objectMeta.Name)
+
+	if err := injectIntoSpec(&i.params, spec); err != nil {
+		return err
+	}
+
+	addAnnotation(objectMeta, i.params.Version)
+
+	// templated annotation to avoid double-injection
+	if templateObjectMeta != nil {
+		addAnnotation(templateObjectMeta, i.params.Version)
+	}
+
+	return nil
+}
+
+func (i *Initializer) createTwoWayMergePatch(prev, curr interface{}, dataStruct interface{}) ([]byte, error) {
+	prevData, err := json.Marshal(prev)
+	if err != nil {
+		return nil, err
+	}
+	currData, err := json.Marshal(curr)
+	if err != nil {
+		return nil, err
+	}
+	return strategicpatch.CreateTwoWayMergePatch(prevData, currData, dataStruct)
+}
+
+func (i *Initializer) initializeDeployment(obj interface{}) error {
+	in := obj.(*appsv1beta1.Deployment)
+	if !i.initializerPresent(&in.ObjectMeta) {
+		return nil
+	}
+
+	glog.Infof("Initializing deployment: %s", in.Name)
+
+	o, err := runtime.NewScheme().DeepCopy(in)
+	if err != nil {
+		return err
+	}
+	out := o.(*appsv1beta1.Deployment)
+
+	if err = i.modifyResource(&out.ObjectMeta, &out.Spec.Template.ObjectMeta, &out.Spec.Template.Spec); err != nil {
+		return err
+	}
+	patchBytes, err := i.createTwoWayMergePatch(in, out, appsv1beta1.Deployment{})
+	if err != nil {
+		return err
+	}
+	_, err = i.clientset.AppsV1beta1().Deployments(in.Namespace).
+		Patch(in.Name, types.StrategicMergePatchType, patchBytes)
```

Comment:
Initializers should already be serialized (via ordered pending list). `Patch()` was carried over from Kelsey's initializer tutorial.  He switched from PUT to PATCH [here](https://github.com/kelseyhightower/kubernetes-initializer-tutorial/commit/43c68cee0a08c8ebdcc40697aa0ebf275b64c195#diff-dbaf6084dce27ca6b2f557721fd256b9).